great_expectations/zep/postgres_datasource.py

```diff
@@ -0,0 +1,141 @@
+
+
+@dataclasses.dataclass(frozen=True)
+class ColumnSplitter:
+    method_name: str
+    column_name: str
+    namespace: str
+    template_params: List[str]
+
+
+class TableAsset(DataAsset):
+    def __init__(
+        self,
+        datasource: Datasource,
+        name: str,
+        table_name: str,
+        # TODO(bdirks): Provide a mechanism to add splitters in __init__
+    ):
+        super().__init__(name)
+        self.table_name = table_name
+        self._datasource = datasource
+        self._name = name
+        self._batch_template: Dict[str, Set[str]] = {}
+        self._column_splitter: Optional[ColumnSplitter] = None
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def datasource(self):
+        return self._datasource
+
+    def get_batch_request(
+        self, options: Optional[BatchRequestOptions] = None
+    ) -> BatchRequest:
+        return BatchRequest(
+            datasource_name=self.datasource.name,
+            data_asset_name=self.name,
+            options=options or {},
+        )
+
+    # This asset type will support a variety of splitters
+    def add_year_and_month_splitter(self, column_name: str, namespace: str = ""):
+        self._column_splitter = ColumnSplitter(
+            method_name="split_on_year_and_month",
+            column_name=column_name,
+            namespace=namespace,
+            template_params=["year", "month"],
+        )
+        return self
+
+
+class PostgresDatasource(Datasource):
+    asset_types = [TableAsset]
+
+    def __init__(self, name: str, connection_str: str):
+        self.name = name
+        self.execution_engine = SqlAlchemyExecutionEngine(
+            connection_string=connection_str
+        )
+        self.assets: Dict[str, DataAsset] = {}
+
+    def add_table_asset(self, name: str, table_name: str) -> TableAsset:
+        asset = TableAsset(datasource=self, name=name, table_name=table_name)
+        self.assets[name] = asset
+        return asset
+
+    # When we have multiple types of DataAssets on a datasource, the batch_request argument will be a Union type.
+    # To differentiate we could use single dispatch or use an if/else (note pattern matching doesn't appear until
+    # python 3.10)
+    def get_batch_list_from_batch_request(
+        self, batch_request: BatchRequest
+    ) -> List[Batch]:
+        # We translate the batch_request into a BatchSpec to hook into GX core.
+        # NOTE: We only produce 1 batch right now
+        data_asset = self.get_asset(batch_request.data_asset_name)
+
+        # We look at the splitters on the data asset and verify that the passed in batch request provides the
+        # correct arguments to specify the batch
+        batch_spec_kwargs = {
+            "type": "table",
+            "data_asset_name": data_asset.name,
+            "table_name": data_asset.table_name,
+            "batch_identifiers": {},
+        }
+        if column_splitter := data_asset._column_splitter:
```

Comment:
Is this allowed in python 3.7?

Reply:
No, but given that this is going to be a brand new workflow, I don't think it's unreasonable to require python 3.8.